README.md
# Boo

Boo is a discord bot brings joy to people.

## What's new in version 1.0.0?

- Whole new commands.
- Add data through developer tools.
- Everyday news.
- Trading.
- AI (I think so).

### Commands

- `$hi` | `$hello`: This one is the AI thing, try this!
- `$spell` | `$spelling`: Test yourself with Vietnamese/English spellings.
- `$slap`: Slap someone you hate
- `$buy`: Buy some items at the market
- `$me`: Display user's profile.

### Developer tools

- `$help-dev`: Display help menu for developer.
- `$userup`: Update users' data (mainly new users).
- `$emoup`: Update server's emojis, usually after you import new emojis to the server. We need to make sure the database have the latest emojis.
- `$add_bless`: Add new conversations to the database (remember to follow exactly the same as the instruction in help menu), note : '\n' is a break-line. 
- `$load_bless`: Load all the conversations to make sure the data in the database is the latest.
- `$show_bless`: Display the last 5 conversations to check if our database is the latest. Also, remember to follow the instruction.
- `$add_spell`: Add new spelling to database. This one is complicated, be careful while entering the arguments (you have to follow exactly the same as the instruction).

-------------------------------------------------
_INSPIRATION FROM RDVN_


data/release_notes
**RELEASE NOTES   |   [`21 August, 2019`]**

**Boo**   `[NEW BOT]`
Boo is a discord bot brings joy to people

**What’s new in version 1.0.0?**
•   Whole new commands.
•   Add data through developer tools.
•   Everyday news.
•   Trading.
•   AI (I think so).

**Commands**     `[NEW]`
•   `$help`: display help menu
•   `$hi` | `$hello`: This one is the AI thing, try this!
•   `$spell` | `$spelling`: Test yourself with Vietnamese/English spellings.
•   `$slap`: Slap someone you hate.
•   `$buy`: Buy some items at the market.
•   `$me`: Display user's profile.

**Developer Tools**     `[NEW]`
[!] Read this carefully if you're interested in helping me import new data to the existing database, then ask me to provide you the `staff` role to access these tools.
•   `$help-dev`: Display help menu for developer.
•   `$emoup`: Update server's emojis, usually after you import new emojis to the server. We need to make sure the database have the latest emojis.
•   `$add_bless`: Add new conversations to the database (remember to follow exactly the same as the instruction in help menu), note : '\n' is a break-line. 
•   `$load_bless`: Load all the conversations to make sure the data in the database is the latest.
•   `$show_bless`: Display the last 5 conversations to check if our database is the latest. Also, remember to follow the instruction.
•   `$add_spell`: Add new spelling to database. This one is complicated, be careful while entering the arguments (you have to follow exactly the same as the instruction).

If Boo has any bug, fire away in <#604030221926268939>! Because this is part of my practice so I'm sorry if this release note annoys you (probably by its length :D) .
That's it! Enjoy your time with Boo.

_INSPIRATION FROM RDVN_


first_bot.py
from datetime import datetime
import discord
import feedparser
from log import get_logger
import time
import pprint
import random

pp = pprint.PrettyPrinter(indent=4)

client = discord.Client()
TOKEN = 'NTg0MjkyMzU1MTM4NTE5MDUz.XPLH2Q.YD_OAt0xO_vzoZhdjxq875rKtgU'
logger, LOG_LEVEL, stream = get_logger()
logger.setLevel(LOG_LEVEL)
logger.addHandler(stream)

url = 'https://vnexpress.net/rss/tin-moi-nhat.rss'

@client.event
async def on_ready():
    logger.info('We have logged in as {}'.format(client.user))
    games = ["with the leaves", "with fire", "with you"]
    game = discord.Game(name=random.choice(games))
    await client.change_presence(status=discord.Status.online, activity=game)

@client.event
async def on_message(message):
    if message.author == client.user:
        return

    # Show help message
    if message.content.startswith('$help'):
        emojis = client.emojis
        kitty = str(emojis[0])
        competition = str(emojis[1])
        embed=discord.Embed(title=kitty + " **Boo - The Happy Virus**",
                            description="Hi I'm Boo and I was assigned a mission to cheer you up",
                            colour=0xfef249)
        # commands
        embed.add_field(name=':hand_splayed: $hello', value="`I'll molasses\ninto your ear`", inline=True)
        embed.add_field(name=':newspaper: $news', value="`I'll get you\nthe latest news`", inline=True)
        embed.add_field(name=competition+' $slap', value="`Have a bad day?\nTell me who you\nwant to slap`", inline=True)
        # footer
        embed.set_footer(text="You're beautiful no matter what ❤️")
        logger.info('Sending help message to {}'.format(message.channel))
        await message.channel.send(embed=embed)

    # Say Hello
    if message.content.startswith('$hello'):
        logger.info('Saying hello')
        current_time = datetime.now()
        morning = [
            'Roses are red\nViolets are blue\nSugar is sweet\n'
            'And so are you\n\nHave a nice day ^^',
            'Hề lô, gụt mó ninh',
            'Cà hê bỏ muối bỏ đường\nBỏ gì cũng được xin em đừng bỏ anh\n:kissing_closed_eyes:',
        ]
        lunch = [
            'Trưa rồi đi ăn trưa đi',
            'Cơm gà ngon đó, đi ăn đi',
            'Ăn đi rồi rửa mặt cho tỉnh táo quay lại làm việc nèk'
        ]
        afternoon = [
            'Chiều rồi đi học bài đi',
            'Không đi học thì chơi đàn đi',
            'Giặt đồ chưa? Đi giặt nhanh',
            'Tranh thủ ngủ trưa xíu đi, sau này không được ngủ đâu'
        ]
        evening = [
            'Đi ăn chiều đi tối rồi.',
            'Hôm nay, chị đi học hay ở nhà?\nCó gì thú vị kể em nghe',
            'Dọn nhà ik, nhà hơi dơ rồi đó',
            'Đi học dzui ko?'
        ]
        late_night = [
            'Giờ này giờ nào rồi sao chưa đi ngủ ???',
            '0 giờ rồi em ngủ đi em\nHơi đâu mà lo kế sinh nhaiii',
            'Mai có đi học không? Sao con mắt còn thao láo z ?',
            'Mai không đi học thì cũng nên đi ngủ sớm đi chị\nYêu chị gất nhiều ❤️',
            'Ông trời ơi ngó xuống mà coi h này nó còn thức nè\nNgủ đi má :mad:',
            'Khuya rồi mà chưa ngủ? Chơi bê đê ko?',
            'Giờ này chưa ngủ hả?\nNghe chuyện ma ko?',
            'Bây dờ, 1 là đi ngủ, 2 là đưa đít đây ?',
            'Ngủ đi chị, thức khuya coi chừng bị bê đê đóa',
        ]
        # 7 <= hour < 10
        if current_time.hour >= 7 and current_time.hour < 10:
            logger.info('It\'s morning so I\'m getting morning responses')
            response = random.choice(morning)
        # 10 <= hour < 13
        elif current_time.hour >= 10 and current_time.hour < 13:
            logger.info('It\'s lunch time !!!')
            response = random.choice(lunch)
        # 13 <= hour < 17
        elif current_time.hour >= 13 and current_time.hour < 17:
            logger.info('It\'s afternoon, doing something helpful')
            response = random.choice(afternoon)
        # 17 <= hour < 22
        elif current_time.hour >= 17 and current_time.hour < 22:
            logger.info('It\'s evening, nothing left to say')
            response = random.choice(evening)
        # 22 <= hour <= 23 or 0 <= hour < 7 
        elif ((current_time.hour >= 22 and current_time.hour <= 23)
                or (current_time.hour >= 0 and current_time.hour < 7)):
            logger.info('It\'s late_night, going to sleep')
            response = random.choice(late_night)
        await message.channel.send(response)
    
    # Get all the users of the server
    if message.content.startswith('$user'):
        logger.info('Getting users of the server')
        await message.channel.send(client.users)

    # Send the latest news from vnexpress to channel
    # báo-mới-mỗi-ngày
    if message.content.startswith('$news'):
        # get the latest news on vnexpress
        feed = feedparser.parse(url)
        # get channel báo-mới-mỗi-ngày
        news_channel = client.get_channel(602426911469207552)
        logger.info('Successfully get information from url')

        for entryNo in range(5):
            # e = discord.Embed()
            # get the title of the article
            title = feed.entries[entryNo].title
            # summary = feed.entries[entryNo].summary
            # get the news image
            # image_link = summary.split('<')[2].split('"')[1]
            # e.set_image(url=image_link)
            # get the link of the article
            link = feed.entries[entryNo].links[0].href
            logger.info('Sending "{}" to {}'.format(title, news_channel))
            await news_channel.send('**{}**\n{}'.format(title, link))

    # Slap contest


def main():
    try:
        client.run(TOKEN)
    except Exception as e:
        logger.error(e)

if __name__ == '__main__':
    main()


utils/command.py
from datetime import datetime
import discord
import random


def generate_help_message(emojis):    
    embed=discord.Embed(
        title=emojis["kitty"] + " **Boo - The Happy Virus**",
        description="Hi I'm Boo and I was assigned a mission to cheer you up",
        colour=0xfef249
    )
    # commands
    embed.add_field(
        name=':hand_splayed: $hello',
        value="`I'll molasses\ninto your ear`",
        inline=True
    )
    embed.add_field(
        name=emojis["competition"]+' $slap',
        value="`Have a bad day?\nTell me who you\nwant to slap`",
        inline=True
    )
    # footer
    embed.set_footer(text="You're beautiful no matter what ❤️")
    return embed

def generate_hello_message(blessings):
    # +7.00 to get Vietnam clock
    current_hour = datetime.now().hour + 7
    if current_hour >= 24:
        current_hour -= 24
    # 7 <= hour < 10
    if current_hour >= 7 and current_hour < 10:
        response = random.choice(blessings["morning"])
    # 10 <= hour < 13
    elif current_hour >= 10 and current_hour < 13:
        response = random.choice(blessings["lunch"])
    # 13 <= hour < 17
    elif current_hour >= 13 and current_hour < 17:
        response = random.choice(blessings["afternoon"])
    # 17 <= hour < 22
    elif current_hour >= 17 and current_hour < 22:
        response = random.choice(blessings["evening"])
    # 22 <= hour <= 23 or 0 <= hour < 7 
    elif ((current_hour >= 22 and current_hour <= 23)
            or (current_hour >= 0 and current_hour < 7)):
        response = random.choice(blessings["late_night"])
    return response, current_hour


utils/json_parser.py
import json


def parse_json_file(filepath):
    with open(filepath, 'r') as f:
        json_data = f.read()
    return json.loads(json_data)


def dump_json_to_file(data_type, json_data, destination_file, exist=False):
    # get blessings or spelling
    if data_type == 'bless' or data_type == 'spell':
        data = json.dumps(json_data, indent=4)
    # get emoji
    if data_type == 'emoji':
        emoji_dict = {}
        for emoji in json_data:
            emoji_dict[emoji.name] = str(emoji)
        data = json.dumps(emoji_dict, indent=4)
    # get user
    if data_type == 'user':
        if exist:
            data = json.dumps(json_data, indent=4)
        else:
            old_users = parse_json_file('data/server_data/users.json')
            for user in json_data:
                # if user already in data and didn't change his name
                if str(user.id) in old_users\
                and old_users[user.id]["name"] == user.name:
                    continue
                # if the user still in data but change their name
                elif str(user.id) in old_users:
                    old_users[user.id]["name"] = user.name
                else: # if this is a new user
                    user_dict = {}
                    user_dict['name'] = user.name
                    user_dict['bot'] = user.bot
                    user_dict['coins'] = 0
                    user_dict['property'] = {}
                    user_dict['property']['slaps'] = 0
                    old_users[user.id] = user_dict
            data = json.dumps(old_users, indent=4)
    # write data to destination file
    with open(destination_file, 'w+') as f:
        f.write(data)


utils/tools.py
from datetime import datetime
import discord
import random
from string import ascii_uppercase


def embed_message(author, color, name, value, thumbnail=False, url=''):
    embed=discord.Embed(colour=color)
    embed.add_field(name=name, value=author+value)
    if thumbnail: embed.set_thumbnail(url=url)
    return embed


def generate_spelling_result(lang, custom_answers, random_question, author,
                             right_answer, meaning, example, kind, pronounce):
    """
    Return embed to send message to the channel

    :param list custom_answers: the answers of the question
    :param str random_question: the question as well as the answer
    :param str author: the author who run the command
    :param str right_answer: the right answer in unicode form
    :param str meaning: meaning of the question
    :param str example: use case of the question
    :param str kind: type of the answer the author chose

    :example:

    >>> custom_answers = ['abc', 'def']
    >>> random_question = 'abc'
    >>> author = <@:123456789098>
    >>> right_answer = 'A'
    >>> kind = 'wrong'
    >>> embed = generate_spelling_result(custom_answers, random_question, author, right_answer, kind)
    >>> print(embed)
    Vietnamese spelling quick question in 10s
    Which one is correct?
    **`[ A ]:` abc**
    ~~`[ B ]:` def~~

    <@:123456789098>, nope, it's A
    """
    if lang == 'vn':
        heading = "Vietnamese"
    elif lang == 'en':
        heading = "English"
    heading += " spelling quick question in 10s\nWhich one is correct?"
    i, result = 0, ''
    for answer in custom_answers:
        if answer == random_question:
            result +=  '**`[ {} ]:` {}\n**'.format(ascii_uppercase[i], answer)
            i += 1
        else:
            result +=  '~~`[ {} ]:` {}\n~~'.format(ascii_uppercase[i], answer)
            i += 1
    def _generate_embed(color):
        embed=discord.Embed(colour=color)
        value = '**Pronounce**: {}\n{}'.format(pronounce, result) if pronounce else result
        embed.add_field(name=heading, value=value, inline=False)
        embed.add_field(name="Meaning", value=meaning, inline=False)
        embed.add_field(name="Example",value=example, inline=False)
        return embed
    if kind == 'wrong':
        result += '\n{}, nope, it\'s {}\n'.format(author, right_answer)
        embed = _generate_embed(0xfa031c)
    elif kind == 'right':
        result += '\n{}, you are absolutely right!\n'.format(author)
        embed = _generate_embed(0x09e30d)
    elif kind == 'time-out':
        result += '\n{}, time out!, the answer is {}\n'.format(author, right_answer)
        embed = _generate_embed(0xfa031c)
    return embed


def generate_help_message(author, message, emojis):
    if message.endswith('dev') and 602501937685987378 \
        in [role.id for role in author.roles]:
        embed=discord.Embed(
            title=emojis["kitty"] + " **Boo - The Happy Virus**",
            description="A help menu for dev only",
            colour=0xfef249
        )
        # commands
        embed.add_field(
            name='`$emoup`',
            value='`Update server\'s\nemojis`',
            inline=True
        )
        embed.add_field(
            name="`$load_bless`",
            value="`Load blessings\nin case blessings\nwere updated\nbut never"
                  " loaded`",
            inline=True           
        )
        embed.add_field(
            name='`$add_bless`',
            value="`Add new bless\n',,': '\\n'\n[time] [new_bless]`",
            inline=True
        )
        embed.add_field(
            name='`$show_bless`',
            value="`Show bless on time\n[time]: morning, lunch, afternoon, "
                  "evening, late_night`"
        )
        embed.add_field(
            name='`$add_spell`',
            value="`Add new spelling\n\"[lang]\" \"[question]\" "
                  "\"[answer1,answer2...]\" \"[meaning]\" \"[example]\" "
                  "\"<pronounce>\"`"
        )
        # footer
        embed.set_footer(text="Dev tool | {}".format(datetime.now()))
    else:    
        embed=discord.Embed(
            title=emojis["kitty"] + " **Boo - The Happy Virus**",
            description="Hi I'm Boo and I was assigned a " \
                        "mission to cheer you up",
            colour=0xfef249
        )
        # commands
        embed.add_field(
            name=':hand_splayed: $hello | $hi',
            value="`I'll molasses\ninto your ear`",
            inline=True
        )
        embed.add_field(
            name=emojis["competition"]+' $slap',
            value="`Have a bad day?\nTell me who you\nwant to slap`",
            inline=True
        )
        embed.add_field(
            name=':pencil: $spell | $spelling',
            value="`Check your spellings\n$spell vn | $spell en`",
            inline=True
        )
        # footer
        embed.set_footer(text="You're beautiful no matter what ❤️")
    return embed

def generate_hello_message(blessings):
    # +7.00 to get Vietnam clock
    current_hour = datetime.now().hour + 7
    if current_hour >= 24:
        current_hour -= 24
    # 7 <= hour < 10
    if current_hour >= 7 and current_hour < 10:
        response = random.choice(blessings["morning"])
    # 10 <= hour < 13
    elif current_hour >= 10 and current_hour < 13:
        response = random.choice(blessings["lunch"])
    # 13 <= hour < 17
    elif current_hour >= 13 and current_hour < 17:
        response = random.choice(blessings["afternoon"])
    # 17 <= hour < 22
    elif current_hour >= 17 and current_hour < 22:
        response = random.choice(blessings["evening"])
    # 22 <= hour <= 23 or 0 <= hour < 7 
    elif ((current_hour >= 22 and current_hour <= 23)
            or (current_hour >= 0 and current_hour < 7)):
        response = random.choice(blessings["late_night"])
    return response, current_hour
